doc/fundamentals/spouts.md

```diff
@@ -0,0 +1,291 @@
+from the environment
+or command-line parameters. 
+That way you can connect to new data sources
+using the same container code
+without recompiling
+by just setting appropriate parameters
+in the pipeline specification.
+```Go
+func main() {
+	// Get the connection info from the ENV vars
+	host := os.Getenv("HOST")
+	port := os.Getenv("PORT")
+	topic := os.Getenv("TOPIC")
+```
+
+### Connect to the streaming data source
+
+We're creating an object
+that can be used to read from our data source.
+
+That `defer` statement is the Go way to guarantee
+that the open file will be closed
+after `main()` returns.
+That is, 
+`reader.Close()` won't be executed
+until after `main()` is finished.
+(This is a common idiom in Go;
+to `defer` the close of a resource
+right after you open it.)
+
+
+```Go
+	// And create a new kafka reader
+	reader := kafka.NewReader(kafka.ReaderConfig{
+		Brokers:  []string{host + ":" + port},
+		Topic:    topic,
+		MinBytes: 10e1,
+		MaxBytes: 10e6,
+	})
+	defer reader.Close()
+```
+
+### Open /pfs/out for writing
+
+We're opening the named pipe`/pfs/out` for writing
+so we can send it a `tar` archive stream
+with the file
+we want to output.
+Note that the named pipe has to be opened
+with write-only permissions.
+
+
+```Go
+    // Open the /pfs/out pipe with write only permissons (the pachyderm spout will be reading at the other end of this)
+	// Note: it won't work if you try to open this with read, or read/write permissions
+	out, err := os.OpenFile("/pfs/out", os.O_WRONLY, 0644)
+	if err != nil {
+		panic(err)
+	}
+	defer out.Close()
+```
+
+### Write the outer file loop
+Here we open a tar stream 
+into the directory we opened above,
+so that Pachyderm can place files
+into the output repo.
+For clarity's sake,
+we'll omit the message-processing loop
+inside this file loop.
+The `err` variable is used
+in the message-processing loop
+for errors reading from the stream
+and writing to the directory.
+The stream is opened 
+at the top of the loop
+and should be closed 
+at the bottom.
+In this case, 
+it'll be closed 
+after a message is processed.
+
+```Go
+	// this is the file loop
+	for {
+		if err := func() error {
+			tw := tar.NewWriter(out)
+			defer tw.Close()
+			// this is the message loop
+			for {
+
+            // ...omitted
+            
+			}
+		}(); err != nil {
+			panic(err)
+		}
+	}
+```
+
+### Create the message processing loop
+
+If you have trouble following this Go code,
+just read the text
+to get an idea of what you need to do.
+
+First, 
+we read a message
+from our Kafka queue.
+Note the use of a 5-second timeout
+on the read.
+
+```Go
+			// this is the message loop
+            for {
+				// read a message
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				m, err := reader.ReadMessage(ctx)
+				if err != nil {
+					return nil
+                }
+```
+
+Then we create a filename
+and write that to the `tar` stream
```

Comment:
maybe worth pointing out that you're not just writing the filename, you're creating the tar header for the file. And importantly, the header needs to include the file size as well.

Reply:
That's a very good point, thank you.  It's essentially giving all the file metadata to the named pipe so that Pachyderm knows  how to handle it.